### ISSUE TITLE
How to use mxnet for image segmentation training?

### ISSUE THREAD

**Issue by tornadomeet** [dmlc/mxnet#337]
In image segmentation, if one image has N pixpels, the number of labels is also N(not one),  so the Softmax operator in mxnet can't handle it(just my opinion); 

I want solve it through:

1. add a new operator(egs, softmaxseg-inl.h, softmaxseg.cu, softmaxseg.cc) in src/operator directory, SoftmaxSeg is used for image segmentation in forward, backword, calc loss..

2. change the format of image list file, like this :  integer_image_index  \t  label.jpg  \t  data.jpg  (each pixel in label.jpg stands for its class label).

3. change the code of iter_image_recordio.cc, so class ImageLabelMap can read label.jpg  and store the value in label_;

    I am a beginner of mxnet, How should we use mxnet for image segmentation training? can the above way solve it? or is there a better solution?

    Give some advise, thanks.

**Comment by pluskid** [comment]
The Caffe / Mocha.jl way of handling this is to allow "multiple dimensions" in the softmax loss layer. For example, the label could be (using Python's row-based ordering) `N-by-1-by-P`, while the predictions will be `N-by-K-by-P` where `N` is number of samples in mini-batch, `K` is number of classes, `P` could be interpreted as number of positions / pixels. In general, the prediction could be any ND tensor, and label will be a tensor with corresponding shape, except one of the dimension is a singleton dimension (of size 1). For example, pixel-wise prediction would be

* Label: `N-by-1-by-H-by-W`
* Prediction: `N-by-K-by-H-by-W`

In terms of image segmentation, I think this [CRF-as-RNN Caffe Code](https://github.com/torrvision/crfasrnn) might be a very nice one to incorporate into MXNet and serving as a cool demo. I might consider doing it when I squeeze out some time but it seems not very recently. So @tornadomeet if you are working in segmentation you could probably add this.

**Comment by pluskid** [comment]
BTW: just as a side comment since our current `Softmax` operator is being discussed. It is a bit confusing at first when I was using the `Softmax` operator. The `forward` operation behave exactly like a softmax, but in the backward operation, it becomes a *softmax with multiclass logistic loss*. I guess this might be due to efficiency consideration or code re-use, but it leads to some inconvenience / inconsistency.

* The `Softmax` operator needs both `data` and `label` as arguments, though if people only do prediction, only the `data` is needed.
* There is no objective function value of the logistic loss computed. Though it does not really affect the learning, and especially we are running for a fixed number of iterations, we do not need to rely on it for stop condition, either. But the correct objective function value serves important role for debugging. For example, when researcher are writing their own layers, looking at how the objective function changing at a finer scale might be very helpful for testing whether it is implemented correctly, and if not where might be some issues.
* People might need to use some other losses on top of the `Softmax` probabilistic output. Caffe / Mocha.jl currently have a `Softmax` layer that does just softmax, and then a `SoftmaxLoss` layer which combines softmax and logistic loss. I think this might be a viable alternative way.

The followings are just some minor thoughts on the general design:

* Currently the `label` arguments for `Softmax` is an `implicit` arguments. If I understand correctly, there is no way to (like `data`) construct a `Variable` as the `label` and compose them like this
```
data = mx.symbol.Variable('data-foo')
label = mx.symbol.Variable('label-foo')
net = mx.symbol.Softmax(data=data, label=label, name='out')
```
This essentially allows us to *rename* the label variable to whatever we want, as in the `data` case. Then when the user construct `DataIter`, he/she can specify what names (`data-foo` and `label-foo` in this case) the `DataIter` is providing (this is also the current design used in MXNet.jl). So when we train the network with this `DataIter` we actually knows which ones are data which ones are labels. The current way (automatically deciding based on a `data` or `label` postfix) is nice, but it might get confused to figure out the exact correspondence in multiple-input multiple-output case (RNN case, not image segmentation case). 
* Since the loss functions are only used during training, we might use the following convention: when defining the architecture, the user only defines the network up to the output layer. A loss layer is provided when the user calls `fit`, and the network is composed with the loss criterion on the fly. But otherwise, the loss is not a part of the architecture and the semantic for doing `predict` with the network is clearer. For example, [Lasagne](http://lasagne.readthedocs.org/en/latest/), a light-weighted framework based on Theano has this design.

**Comment by tqchen** [comment]
@pluskid I agree with your idea on softmax should be solely softmax transformation. 
And softmax-loss should be used for loss function. Do you want to take a stab on the refactor?

**Comment by tqchen** [comment]
The care need to make a truely valid softmax operator, though. Because being able to propback the right gradient for any composition after softmax requires the log-probability instead of probability for numerical stability. A better approach might be take a log-softmax operator to allow arbitrary composition. Or simply restrict softmax to only be able to composed with SoftmaxLoss.

The current decision was made to make things work in a good way for restricted case, which should be changed.

**Comment by antinucleon** [comment]
@tqchen We can do refactor soon after we fix RNN stuff, because 2 reason: 1. Capable with CuDNN 2. Better support different loss on Softmax

**Comment by winstywang** [comment]
Hi, all

  Thanks are quite a lot of work need to be done for segmentation task. The multi softmax loss for segmentation is only one issue. However, there are several other ops need to be implemented:

1. The UpSampling layer used in FCN
2. IO part for image segmentation

The most concerned part is the UpSampling layer. It may not easy to implement in current mshadow framework.

**Comment by antinucleon** [comment]
@winstywang 
1. We can enable calling CUDA Kernel directly in OP instead of writing mshadow OP
2. It is not hard to write Python IO directly instead of writing special C++ IO for segmentation

**Comment by pluskid** [comment]
@tqchen Sure I'm glad to help on this but let me roll MXNet.jl in a good shape first. We might need to have somewhere an official TODO items for the whole project so that we do not forget things.

@winstywang I heard from a friend working on image segmentation saying that 

> though the up-sampling layer could be formulated in various ways, using, for example, fancy un-convolution, etc. In the end, a naive up-sampling by image-resizing-kind of operation just works fine. Even those complicated up-sampling layers are initialized in this way, and if you look at the learned filters in those layers, they are not really far from the initialization, which is simple up-resizing.

I have not personally tried to implement and verify this statement. But I think it might be worth to keep in mind. If this is true, it will definitely save everybody a lot of extra work.

**Comment by pluskid** [comment]
@winstywang What do you mean by the IO part? If our softmax is extended to Caffe-like that support evaluating at each pixels, then the whole pipeline is still single-input single-output (though multi-dimensional). It seems the current pipeline fits, or is the current image-records-IO only support single value labels?

**Comment by pluskid** [comment]
@tqchen I agree that combined softmax and logistic loss is much more numerically stable. The compromise in Caffe/Mocha.jl is to provide a `SoftmaxLoss` layer that is essentially combined softmax and logistic loss layer, but do (backward) computation together. It is more efficient *and* numerically stable. But meanwhile, also offer softmax layer and logistic regression layer separately in case people need to do something else. Actually, a while ago, I was working on a project that use a different loss other than the logistic on the softmax output.

In fact, doing them separately might not be that bad than we think in practice, if you do softmax carefully. I did some experiments quite long time ago (scroll down the [page for the last figure](http://freemind.pluskid.org/machine-learning/softmax-vs-softmax-loss-numerical-stability/)). It seems that the discrepancy is within a reasonable range for relatively bounded inputs even for `Float32`. As far as I know, Theano just [provide softmax and logistic loss separately](http://deeplearning.net/software/theano/library/tensor/nnet/nnet.html), though I do not know if during the compilation stage this actually get optimized into a single `softmax-loss` optimization. If we are still concerned here, the option adopted by Torch might also be possible: they have a `LogSoftmax` layer. :smile:

**Comment by tornadomeet** [comment]
Thansk all.
@pluskid Yes, i want to do some experiments on segmentation, using [deeplab code](https://bitbucket.org/deeplab/deeplab-public/), and [CRF-as-RNN code](https://github.com/torrvision/crfasrnn), If I have the ability to complete it, i'll share it, but First let me be familar with mxnet . ^^

**Comment by tqchen** [comment]
The multi label softmax is merged in #387

**Comment by tornadomeet** [comment]
@tqchen ,Yeah, i see it! thanks

**Comment by futurely** [comment]
Torch's upsampling implementation is [here](https://github.com/torch/nn/blob/master/SpatialUpSamplingNearest.lua).

**Comment by futurely** [comment]
The current state-of-the-art on the [VOC 2012 leaderboard of segmentation competition](http://host.robots.ox.ac.uk:8080/leaderboard/displaylb.php?challengeid=11&compid=6) is deep parsing network [1]. Not implemented operations include local convolution, channels block min pooling, element-wise log and element-wise exponent.

[1] Ziwei Liu, Xiaoxiao Li, Ping Luo, Chen Change Loy, and Xiaoou Tang. Semantic Image Segmentation via Deep Parsing Network. ICCV 2015.

**Comment by futurely** [comment]
@HyeonwooNoh's entry in the [PASVOC VOC segmentation](http://host.robots.ox.ac.uk:8080/leaderboard/displaylb.php?challengeid=11&compid=6) benchmark 	[POSTECH_DeconvNet_CRF_VOC](http://host.robots.ox.ac.uk:8080/leaderboard/displaylb.php?challengeid=11&compid=6#KEY_POSTECH_DeconvNet_CRF_VOC) achieved 74.8% average precision. In [his implementation](https://github.com/HyeonwooNoh/caffe/commits/new_data_layer), pixel wise classification loss [RedSoftmaxWithLossLayer](RedSoftmaxWithLossLayer), [EltwiseAccuracyLayer](https://github.com/HyeonwooNoh/caffe/blob/new_data_layer/src/caffe/layers/eltwise_accuracy_layer.cpp) and [RedAccuracyLayer](https://github.com/HyeonwooNoh/caffe/blob/new_data_layer/src/caffe/layers/red_accuracy_layer.cpp) are related to segmentation. The C++ [seg data layers](https://github.com/HyeonwooNoh/caffe/blob/master/include/caffe/data_layers.hpp) are better replaced by [Python layers](http://mxnet.readthedocs.org/en/latest/tutorial/new_op_howto.html) like https://github.com/BVLC/caffe/issues/1698#issuecomment-144269770 and https://github.com/dmlc/mxnet/issues/554#issuecomment-156230633 which would be easier to implement.

[1] Hyeonwoo Noh, Seunghoon Hong, Bohyung Han. [Learning Deconvolution Network for Semantic Segmentation](http://cvlab.postech.ac.kr/research/deconvnet/). ICCV 2015.

**Comment by HyeonwooNoh** [comment]
@futurely 
EltwiseAccuracyLayer is used to compute accuracy for segmentation, but
RedSoftmaxWithLossLayer is not used for segmentation (it's implemented for another purpose).
For segmentation, SoftmaxLoss layer should be refered.

**Comment by winstywang** [comment]
wow, the authors are here :) 
I will work on the segmentation task soon

**Comment by futurely** [comment]
https://github.com/HyeonwooNoh/DeconvNet/blob/master/training/001_stage_1/stage_1_train.prototxt

**Comment by playerkk** [comment]
Is that possible to assign different weights to different ground-truth labels? The weight can be set as 1/f_i, where f_i is the frequency of the i-th label in the training set, to deal with the unbalanced number of training samples of different classes.

**Comment by futurely** [comment]
In "Learning Deconvolution Network for Semantic Segmentation", "for both datasets, we maintain the balance for the number of examples across classes by adding redundant examples for the classes with limited number of examples."

**Comment by futurely** [comment]
In more [traditional](http://scikit-learn.org/stable/auto_examples/linear_model/plot_sgd_weighted_samples.html) machine learning algorithms, [weighted sample](http://scikit-learn.org/stable/auto_examples/svm/plot_weighted_samples.html) are not unusual. But long tail samples sometimes hurt the performance in DL experiments [2]. It's doubtful whether duplicating them will be of any help.

[2] Erjin Zhou, Zhimin Cao, Qi Yin: Naive-Deep Face Recognition: Touching the Limit of LFW Benchmark or Not? CoRR abs/1501.04690 (2015).

**Comment by playerkk** [comment]
@futurely Please take a look at Eq.(1) of http://www.cv-foundation.org/openaccess/content_cvpr_2015/papers/Mostajabi_Feedforward_Semantic_Segmentation_2015_CVPR_paper.pdf

**Comment by futurely** [comment]
As you can see on the PASCAL VOC 2012 segmentation [leaderboard](http://host.robots.ox.ac.uk:8080/leaderboard/displaylb.php?challengeid=11&compid=6#KEY_POSTECH_DeconvNet_CRF_VOC), most top performing submissions are using the Microsoft COCO dataset to further boost performance. The observation is very similar with what was discovered in "Naive-Deep Face Recognition": big data is more effective than complicated algorithms.

**Comment by winstywang** [comment]
I will work on segmentation task from next week. I think the first step is to implement the basic FCN model instead of the second model.

**Comment by futurely** [comment]
@playerkk class weighting is [implemented](https://github.com/alexgkendall/caffe-segnet/search?utf8=%E2%9C%93&q=class_weighting) by the authors of [3][4]. But the experimental result on the PASCAL VOC 2012 segmentation dataset is not very competitive with DeconvNet [1] and CRF-RNN [5] although its network architecture seems to be very similar with DeconvNet. It's would be instructive to get some explanations from @alexgkendall.

[3] Alex Kendall, Vijay Badrinarayanan and Roberto Cipolla "Bayesian SegNet: Model Uncertainty in Deep Convolutional Encoder-Decoder Architectures for Scene Understanding." arXiv preprint arXiv:1511.02680, 2015. http://arxiv.org/abs/1511.02680
[4] Vijay Badrinarayanan, Alex Kendall and Roberto Cipolla "SegNet: A Deep Convolutional Encoder-Decoder Architecture for Image Segmentation." arXiv preprint arXiv:1511.00561, 2015. http://arxiv.org/abs/1511.00561
[5] Shuai Zheng, Sadeep Jayasumana, Bernardino Romera-Paredes, Vibhav Vineet, Zhizhong Su, Dalong Du, Chang Huang and Philip H.S. Torr, Conditional Random Fields as Recurrent Neural Networks. IEEE International Conference on Computer Vision (ICCV), 2015.

**Comment by alexgkendall** [comment]
Hi - Our work [3][4] differs from [1] in class weighting as you pointed out. Another major difference is the much more efficient parametrisation in SegNet [4], which is an order of magnitude faster to run, and can be trained end to end in one step. I believe DeConvNet [1] uses stage wise training and multiple region proposals in inference time.

We found class weighting to be very important for scene understanding tasks. If you are more interested in datasets such as [SUN](http://mi.eng.cam.ac.uk/research/projects/VideoRec/CamVid/) or [CamVid](http://rgbd.cs.princeton.edu/) then I'd recommend implementing it. Cheers.

**Comment by playerkk** [comment]
Thanks all for your reply.

I am working on predicting pixel-wise labels, not necessarily image segmentation. The distributions of labels are highly unbalanced. Experimental results based on MatConvNet indicate that weighted loss is quite helpful. MatConvNet, however, is very slow. So I am considering to switch using other packages.

**Comment by alexgkendall** [comment]
Cool, SegNet can do that and it is built on Caffe which is pretty fast - you might find [this tutorial](http://mi.eng.cam.ac.uk/projects/segnet/tutorial.html) helpful.

**Comment by futurely** [comment]
Once https://github.com/dmlc/mxnet/issues/640 is completed, most of the published image segmentation models that are based on FCN can be replicated without much difficulty in MXNet.

**Comment by futurely** [comment]
Actually, convolutional autoencoder networks don't have to use the pooling and unpooling layers. Simply stacking several conv layers and then a few deconv layers together is alright. Just for a little more fun, they can be arbitrarily interleaved!

**Comment by mli** [comment]
@tornadomeet congras you make it work, i'm going to close this issue now. it will be great if you can PR the example back.

to others, the example is available at https://github.com/tornadomeet/mxnet/tree/seg/example/fcn-xs

**Comment by zhangfanqie** [comment]
@tornadomeet Hi,I'm confused about how to make my own training data.Does the im2rec.py support making .rec files when labels are images as well? How to do with its arguments?
Thanks a lot!!!

**Comment by great-thoughts** [comment]
If my data resides in AWS S3,how should I modify the code? passing s3://bucketname/... doesnt seem to work.
